yktracec/src/jitmodbuilder.cc

```diff
@@ -1426,73 +1426,40 @@ class JITModBuilder {
       TraceLoc Loc = InpTrace[Idx];
 
       if (UnmappableRegion *UR = Loc.getUnmappableRegion()) {
-        // The trace entered a region of unmappable foreign code.
-        //
-        // As noted in the mapper and asserted in the JITModBuilder constructor,
-        // there are never two unmappable regions in a row, so the next block
-        // (if there is one) is guaranteed to be mappable. The question is: will
-        // the foreign code be *returning* into mappable code, or will it be
-        // *calling back* deeper into mappable code? The trace decoder keeps
-        // track of the stack effects of foreign code (assuming its control flow
-        // isn't bonkers), so we are able to use that information to decide.
-
-        if (InpTrace.Length() == Idx + 1) {
-          // This unmappable region is the end of the trace, so there's no
-          // need to maintain the stack any more.
-          continue;
-        }
+        continue;
+      }
 
-        assert(!CallStack.curMappableFrame());
-
-        // FIXME: think about what do do in the face of setjmp/longjmp. How
-        // would we find the stack adjustment value for such code?
-        if (UR->StackAdjust < 0) {
-          // The stack got smaller as a result of executing the foreign code,
-          // so we must be returning to mappable code.
-          while (UR->StackAdjust < 0) {
-            // We don't allow foreign code to pop non-foreign frames. That
-            // seems like a sure indiciator of bonkers control flow.
-            assert(!CallStack.curMappableFrame());
+      IRBlock *IB = Loc.getMappedBlock();
+      assert(IB);
+      CurBBIdx = IB->BBIdx;
+
+      auto [F, BB] = getLLVMAOTFuncAndBlock(IB);
+
+      if (CallStack.curMappableFrame() == nullptr) {
+        if (!BB->isEntryBlock()) {
+          // The current block is not an entry block and we have just seen an
```

Comment:
Should we state the property of the IR that allows us to make this deduction: that an entry block can contain no calls (aside from potentially mallocing the shadow stack)?

Reply:
[377beb7](https://github.com/ykjit/yk/pull/911/commits/377beb7322c9de0da57358f2f554fafd68e4a639)

Reply:
One small nitpick.

> This only works when `--yk-no-calls-in-entryblocks` flag is used

I think we should something more like:

> We rely on the `NoCallsOnEntryBlocks` pass to provide the necessary disambiguation for this to work: Since no calls can appear in an entry block... <explain>...

Reply:
[e44242c](https://github.com/ykjit/yk/pull/911/commits/e44242cdc30c940f1d8c5362aa574a46cda52c82)